src/core/reaction_methods/ReactionAlgorithm.cpp

```diff
@@ -564,38 +563,53 @@ ReactionAlgorithm::generate_new_particle_positions(int type, int n_particles) {
       p_id = get_random_p_id(type, random_index);
     }
     drawn_pids.emplace_back(p_id);
-    // store original position
+    // store original state
     auto const &p = get_particle_data(p_id);
-    old_positions.emplace_back(std::pair<int, Utils::Vector3d>{p_id, p.pos()});
-    // write new position
+    old_state.emplace_back(std::tuple<int, Utils::Vector3d, Utils::Vector3d>{
+        p_id, p.pos(), p.v()});
+    // write new position and new velocity
     auto const prefactor = std::sqrt(kT / p.mass());
     auto const new_pos = get_random_position_in_box();
     move_particle(p_id, new_pos, prefactor);
     check_exclusion_range(p_id);
+    if (particle_inside_exclusion_range_touched) {
+      break;
+    }
   }
 
-  return old_positions;
+  return old_state;
 }
 
 /**
- * Performs a global MC move for a particle of the provided type.
+ * Performs a global MC move for particles of a given type.
+ * Particles are selected without replacement.
+ * @param type     Type of particles to move.
+ * @param n_part   Number of particles of that type to move.
+ * @returns true if all moves were accepted.
  */
-bool ReactionAlgorithm::do_global_mc_move_for_particles_of_type(
-    int type, int particle_number_of_type_to_be_changed) {
-  m_tried_configurational_MC_moves += 1;
-  particle_inside_exclusion_range_touched = false;
+bool ReactionAlgorithm::displacement_move_for_particles_of_type(int type,
+                                                                int n_part) {
 
-  auto const particle_number_of_type = number_of_particles_with_type(type);
-  if (particle_number_of_type == 0 or
-      particle_number_of_type_to_be_changed == 0) {
-    // reject
-    return false;
+  if (type < 0) {
+    throw std::domain_error("Parameter 'type_mc' must be >= 0");
+  }
+  if (n_part <= 0) {
+    throw std::domain_error(
+        "Parameter 'particle_number_to_be_changed' must be >= 1");
+  }
+  auto const n_particles_of_type = number_of_particles_with_type(type);
+  if (n_part > n_particles_of_type) {
+    throw std::domain_error(
+        "Parameter 'particle_number_to_be_changed' must be less or equal to "
+        "the number of particles in the system");
   }
```

Comment:
We should spend a second thought on this change, because it will probably break old espresso scripts.
What I was thinking of is a reactive system (e.g. GCMC) where you have initially setup a species that will react with time. At some point it might happen that this species has completely reacted, but one might have used displacement moves in the script to change the configuration. In such a case one would have just rejected the move in the old implementation, but now an error is thrown. An additional, explicit check before the displacement moves would be necessary to prevent this error.
The very same happening is when providing zero, which might have happened when using the current particle number as the number of displacement moves.

Disclaimer: I'm not claiming any of this is good practice, but it used to work and I have written scripts where this change would also break my code.